lib/Feature/FeatureModelParser.cpp

```diff
@@ -233,4 +234,318 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
 // TODO(s9latimm): replace with builder err
 bool FeatureModelXmlParser::verifyFeatureModel() { return parseDoc().get(); }
 
+//===----------------------------------------------------------------------===//
+//                        FeatureModelSxfmParser Class
+//===----------------------------------------------------------------------===//
+
+std::unique_ptr<FeatureModel> FeatureModelSxfmParser::buildFeatureModel() {
+  std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> Doc = parseDoc();
+  if (!Doc) {
+    return nullptr;
+  }
+
+  FMB.init();
+  return parseVm(xmlDocGetRootElement(Doc.get())) ? FMB.buildFeatureModel()
+                                                  : nullptr;
+}
+
+std::unique_ptr<xmlDtd, void (*)(xmlDtdPtr)>
+FeatureModelSxfmParser::createDtd() {
+  std::unique_ptr<xmlDtd, void (*)(xmlDtdPtr)> Dtd(
+      xmlIOParseDTD(nullptr,
+                    xmlParserInputBufferCreateMem(SxfmConstants::DtdRaw.c_str(),
+                                                  SxfmConstants::DtdRaw.length(),
+                                                  XML_CHAR_ENCODING_UTF8),
+                    XML_CHAR_ENCODING_UTF8),
+      xmlFreeDtd);
+  xmlCleanupParser();
+  assert(Dtd && "Failed to parse DTD.");
+  return Dtd;
+}
+
+std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelSxfmParser::parseDoc() {
+  // Initialize the XML parser
+  std::unique_ptr<xmlParserCtxt, void (*)(xmlParserCtxtPtr)> Ctxt(
+      xmlNewParserCtxt(), xmlFreeParserCtxt);
+  // Parse the given model by libxml2
+  std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> Doc(
+      xmlCtxtReadMemory(Ctxt.get(), Sxfm.c_str(), Sxfm.length(), nullptr, nullptr,
+                        XML_PARSE_NOBLANKS),
+      xmlFreeDoc);
+  xmlCleanupParser();
+
+  // In the following, the document is validated.
+  // Therefore, (1) check whether it could be parsed
+  if (Doc && Ctxt->valid) {
+    // (2) validate the sxfm format by using the dtd (document type definition) file
+    if (xmlValidateDtd(&Ctxt->vctxt, Doc.get(), createDtd().get())) {
+      // and (3) check the tree-like structure of the embedded feature model
+      // as well as constraints
+      return Doc;
+    } else {
+      std::cerr << "Failed to validate DTD." << std::endl;
```

Comment:
```suggestion
      llvm::errs() << "Failed to validate DTD.\n";
```
Please try to use llvm streams because they work better together with other llvm string formating etc.